jsonlint.1
.TH jsonlint "1" "December 15, 2023" "" "jsonlint manual"

.SH NAME
jsonlint - JSON/JSONC/JSON5 validator and pretty-printer

.SH SYNOPSIS
.B jsonlint
[-o|--output
.I file
] [-m|--mode
.I
mode
] [-w|--overwrite]
.br
[-p|--pretty] [-k|--check] [-a|--compact] [-t|--trailing-commas]
.br
[-s|--single-quotes] [--escape-slashes] [--escape-unicode]
.br
[-l|--line-break] [-V|--version] [-h|--help]
[
.I [file ...]
]
.RE

.SH DESCRIPTION
jsonlint is a JSON/JSONC/JSON5 validator and pretty-printer.

.SH ARGUMENTS
.B
.IP "file"
read the JSON/JSONC/JSON5 input from a file

.SH OPTIONS
.B
.IP "-o|--output file"
write the JSON output to a file
.B
.IP "-m|--mode mode"
parse in the mode "json", "jsonc", or "json5"
.B
.IP "-w|--overwrite"
overwrite the input file with the formatted output
.B
.IP "-k|--check"
check the syntax only, no output
.B
.IP "-a|--compact"
print error messages on a single line
.B
.IP "-t|--trailing-commas"
insert trailing commas to arrays and objects
.B
.IP "-s|--single-quotes"
format single-quoted instead of double-quoted strings
.B
.IP "--escape-slashes"
escape slashes by by prefixing them with a backslash
.B
.IP "--escape-unicode"
escape multibyte Unicode characters with \u literals
.B
.IP "-l|--line-break"
append a line break to the JSON output
.B
.IP "-p|--pretty"
print the JSON output with line breaks and indented
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE
If no input file is specified, it will be read from standard input.
If multiple files are specified and file overwriting is not enabled,
the files will be only checked and their names printed out.

.SH EXAMPLES
.B
.IP "jsonlint config.json -lpw"
Prettify a file and ensure a trailing line break.
.B
.IP "cat config.json | jsonlint > config2.json"
Minify a file and write the result to a new file.

.SH BUGS
jsonlint's bug tracker is located at <https://github.com/prantlf/v-jsonlint/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/v-jsonlint#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.3.1](https://github.com/prantlf/v-jsonlint/compare/v0.3.0...v0.3.1) (2023-12-15)

### Bug Fixes

* Include man pages in asset archives ([f6fc610](https://github.com/prantlf/v-jsonlint/commit/f6fc6102b733cfc62aa53888c9b4ab173ff36bc1))

## [0.3.0](https://github.com/prantlf/v-jsonlint/compare/v0.2.0...v0.3.0) (2023-12-15)

### Features

* Add the man file ([e97421b](https://github.com/prantlf/v-jsonlint/commit/e97421bfbd842498e490c508cb935c369bab6180))

## [0.2.0](https://github.com/prantlf/v-jsonlint/compare/v0.1.0...v0.2.0) (2023-12-14)

### Features

* Add the linux arm64 support ([04a6011](https://github.com/prantlf/v-jsonlint/commit/04a60118d6761659bc392131cf71887c88abb985))

## [0.1.0](https://github.com/prantlf/v-jsonlint/compare/v0.0.6...v0.1.0) (2023-12-12)

### Features

* Add arm64 build, cross-compile for Windows ([6d83f3c](https://github.com/prantlf/v-jsonlint/commit/6d83f3c5bddeceec3337bcf4913103fffb174429))

## [0.0.6](https://github.com/prantlf/v-jsonlint/compare/v0.0.5...v0.0.6) (2023-10-24)

### Bug Fixes

* Build with updated dependencies ([2d78fb0](https://github.com/prantlf/v-jsonlint/commit/2d78fb08d715b158ec857b9a5cd13e02841e4671))

## [0.0.5](https://github.com/prantlf/v-jsonlint/compare/v0.0.4...v0.0.5) (2023-10-23)

### Bug Fixes

* Update the imprinted version number ([b623809](https://github.com/prantlf/v-jsonlint/commit/b62380929d7c34eb6c7a90a2ae150c340a359d42))

## [0.0.4](https://github.com/prantlf/v-jsonlint/compare/v0.0.3...v0.0.4) (2023-08-07)

### Bug Fixes

* Upgrade to the latest prantlf.cargs ([59e8562](https://github.com/prantlf/v-jsonlint/commit/59e85629ac7a99a5529915eab0e0c179116bcb93))

## [0.0.3](https://github.com/prantlf/v-jsonlint/compare/v0.0.2...v0.0.3) (2023-07-09)

### Features

* Allow to control escaping slashes and unicode characters ([bf6597b](https://github.com/prantlf/v-jsonlint/commit/bf6597b31f23bd40529b0f8ce36d89fb89f46c54))

## [0.0.2](https://github.com/prantlf/v-jsonlint/compare/v0.0.1...v0.0.2) (2023-07-09)

### Features

* Add partial support for JSON5 ([94a287b](https://github.com/prantlf/v-jsonlint/commit/94a287bd9c8e877b558ea1b0c2a0be3a954a6c71))

## 0.0.1 (2023-06-25)

Initial release.


src/jsonlint.v
import os
import prantlf.cargs
import prantlf.json { JsonError, ParseOpts, StringifyOpts, parse, stringify }

const version = '0.3.1'

const usage = 'JSON/JSONC/JSON5 validator and pretty-printer.

Usage: jsonlint [options] [<file> ...]

  <file>                read the JSON/JSONC/JSON5 input from a file

Options:
  -o|--output <file>    write the JSON output to a file
	-m|--mode <mode>      parse in the mode "json", "jsonc", or "json5"
  -w|--overwrite        overwrite the input file with the formatted output
  -k|--check            check the syntax only, no output
  -a|--compact          print error messages on a single line
	-t|--trailing-commas  insert trailing commas to arrays and objects
	-s|--single-quotes    format single-quoted instead of double-quoted strings
	--escape-slashes      escape slashes by by prefixing them with a backslash
	--escape-unicode      escape multibyte Unicode characters with \\u literals
  -l|--line-break       append a line break to the JSON output
  -p|--pretty           print the JSON output with line breaks and indented
  -V|--version          print the version of the executable and exit
  -h|--help             print the usage information and exit

If no input file is specified, it will be read from standard input.
If multiple files are specified and file overwriting is not enabled,
the files will be only checked and their names printed out.

Examples:
  $ jsonlint config.json -lpw
  $ cat config.json | jsonlint > config2.json'

enum Mode {
	json
	jsonc
	json5
}

struct Opts {
mut:
	output          string
	mode            Mode
	overwrite       bool
	check           bool
	compact         bool
	trailing_commas bool
	single_quotes   bool
	escape_slashes  bool
	escape_unicode  bool
	line_break      bool
	pretty          bool
}

fn check_one(file string, names_only bool, opts &Opts) ! {
	input := if file.len > 0 {
		os.read_file(file)!
	} else {
		os.get_raw_lines_joined()
	}

	jsonc := opts.mode == Mode.jsonc
	json5 := opts.mode == Mode.json5
	src := parse(input, ParseOpts{
		ignore_comments: jsonc || json5
		ignore_trailing_commas: jsonc || json5
		allow_single_quotes: json5
	}) or {
		if err is JsonError {
			if file.len > 0 {
				msg := if opts.compact {
					err.reason
				} else {
					err.msg_full()
				}
				eprintln('${file}:${err.line}:${err.column}: ${msg}')
				return
			}
			msg := if opts.compact {
				err.msg()
			} else {
				err.msg_full()
			}
			eprintln(msg)
			return
		}
		eprintln('${file}: ${err.msg()}')
		return
	}

	mut dst := stringify(src, StringifyOpts{
		pretty: opts.pretty
		trailing_commas: opts.trailing_commas
		single_quotes: opts.single_quotes
		escape_slashes: opts.escape_slashes
		escape_unicode: opts.escape_unicode
	})

	if !opts.check {
		if opts.line_break {
			dst += '\n'
		}
		if opts.output.len > 0 {
			os.write_file(opts.output, dst)!
		} else if opts.overwrite && file.len > 0 {
			os.write_file(file, dst)!
		} else if names_only {
			println('${file}: OK')
		} else {
			print(dst)
		}
	}
}

fn check() ! {
	opts, args := cargs.parse[Opts](usage, cargs.Input{ version: version })!

	if args.len > 0 {
		names_only := args.len > 1 && !opts.overwrite
		for arg in args {
			check_one(arg, names_only, opts)!
		}
	} else {
		check_one('', false, opts)!
	}
}

fn main() {
	check() or {
		eprintln(err.msg())
		exit(1)
	}
}
